symtabAPI/src/emitElfStatic-riscv64.C

```diff
@@ -0,0 +1,213 @@
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+char emitElfStatic::getPaddingValue(Region::RegionType) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+void emitElfStatic::cleanupTLSRegionOffsets(map<Region *, LinkMap::AllocPair> &,
+        Region *, Region *)
+{
+    assert(!EMIT_STATIC_ASSERT);
+}
+
+bool emitElfStatic::isGOTRelocation(unsigned long) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+#ifdef OLD_VERSION
+Offset emitElfStatic::getGOTSize(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+#endif
+
+Offset emitElfStatic::getGOTSize(Symtab *, LinkMap &, Offset &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+Offset emitElfStatic::getGOTAlign(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+#ifdef OLD_VERSION
+void emitElfStatic::buildGOT(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+}
+#endif
+
+void emitElfStatic::buildGOT(Symtab *, LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+}
+
+bool emitElfStatic::isConstructorRegion(Region *) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+Offset emitElfStatic::layoutNewCtorRegion(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+bool emitElfStatic::createNewCtorRegion(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+bool emitElfStatic::isDestructorRegion(Region *) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+bool emitElfStatic::isGOTRegion(Region *) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+Offset emitElfStatic::layoutNewDtorRegion(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+bool emitElfStatic::createNewDtorRegion(LinkMap &) {
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+void emitElfStatic::getExcludedSymbolNames(set<string> &) {
+    assert(!EMIT_STATIC_ASSERT);
+}
+
+Offset emitElfStatic::allocStubRegions(LinkMap &, Offset){
+    assert(!EMIT_STATIC_ASSERT);
+    return 0;
+}
+
+bool emitElfStatic::updateTOC(Symtab *, LinkMap &, Offset){
+    assert(!EMIT_STATIC_ASSERT);
+    return false;
+}
+
+inline
+static bool adjustValInRegion(Region *reg, Offset offInReg, Offset addressWidth, int adjust) {
+    Offset newValue;
+    unsigned char *oldValues;
+
+    oldValues = reinterpret_cast<unsigned char *>(reg->getPtrToRawData());
+    memcpy(&newValue, &oldValues[offInReg], addressWidth);
+    newValue += adjust;
+    return reg->patchData(offInReg, &newValue, addressWidth);
+}
+
+bool emitElfUtils::updateRelocation(Symtab *obj, relocationEntry &rel, int library_adjust) {
+    Region *targetRegion = obj->findEnclosingRegion(rel.rel_addr());
+    if (targetRegion == NULL) {
+        rewrite_printf("Failed to find enclosing Region for relocation");
+        return false;
+    }
+    unsigned addressWidth = obj->getAddressWidth();
+    switch (rel.getRelType()) {
+        case R_RISCV_RELATIVE:
+        case R_RISCV_IRELATIVE:
+            rel.setAddend(rel.addend() + library_adjust);
```

Comment:
If `R_RISCV_RELATIVE` corresponds to a `DT_REL` entry, then this will create an addend (with value `library_adjust`) for an entry that logically doesn't have one. Does riscv use `DT_RELA`?

Reply:
That seems to be more the default these days, but I wanted to double-check.